### DISCUSSION TITLE
Can this package be installed as a dev dependency?

### DISCUSSION THREAD

**Discussion by snax4a** [dcastil/tailwind-merge#99] [Help] [answered]
Hi, 
Can this package be installed as a dev dependency?

**Comment by dcastil** [comment, marked as answer]
Yes, any npm package can be installed as devDependency.

**Reply by verheyenkoen** [reply]
Sure, technically you're right but I guess the question is more like: should this be installed as a devDependency? I think if you need a package at runtime (not just at dev/build time), it cannot be in `devDependencies`. So in this case I'd argue tailwind-merge should be in `dependencies`, no?

**Reply by dcastil** [reply]
Yes, you're right, most people should install this as a `dependency`. But where you should install it really depends on the context you're working in.

E.g. when developing a library, it's really important to only have external dependencies in `dependencies`. If a library bundles tailwind-merge as part of itself, it should not be specified as `dependency` but rather a `devDependency`. When building an app that is not depended upon, it's more of a convention what goes where and from a technical standpoint it doesn't matter so much.

**Reply by LewisCowlesMotive** [reply]
For a create-next or create-react project, where it is not intended to be deployed as a library, I wanted to know if I should `npm i -D` or `npm i`. I'm just going to try it out as this has not helped me explicitly know.

transient / bundled dependency -> use runtime / dependencies
direct application dependency -> use devDependencies

I'm sure that was the meaning of what was written, but it's not obvious to me;and the initial answer that packages can be installed in either was quite unhelpful.

I think what was really being asked, is if we install as a runtime dependency in an app; will this be pulled in as a dependency, or will this just be used to build, like https://github.com/jamiebuilds/tailwindcss-animate which (IMO correctly) documents it's a dev dependency.

This stuff is hard, thank you for publishing the code, and sorry if it comes across like I'm telling you off (I'm not); I'm just trying to get to a clear place for the next person that goes looking.

**Reply by dcastil** [reply]
Hey @LewisCowlesMotive! 👋

The short answer is that unless you're building a library it doesn't matter whether you install a package as a dependency or devDependency and you're free to put it wherever you want.

You could put all your packages into dependencies and it wouldn't change the behavior of your Next.js app. In a Next.js app I think most people put tailwind-merge into dependencies, even if it is only in server-rendered components. Server components can still be executed at runtime based on your Next.js config and the API you're using, so it's not necessarily true that tailwind-merge only executes at buildtime. And tailwind-merge can still land in the client bundle if there is a `"use client"` anywhere in the import chain of that component. So dependencies of server components are usually not seen as devDependencies.

**Reply by LewisCowlesMotive** [reply]
Hey @dcastil 👋 

I think for me it can help a number of use-cases to include as much as wont disrupt an app by using `devDependencies` as `npm audit --omit=dev` will ignore package vulnerabilities which only run locally and present less of a risk to end users (pinning and locking should help keep safe-ish).

I think then the rules are:

If this is only being used at build time, and you're not a library author, use devDependencies.
If this is being used at runtime in the end artifact, you must use dependencies or face error?

I'm sure tree shaking could add more nuance to this; but I try to build little decision matricies in my head to help me think more clearly and avoid sloppiness.
